crates/iceberg/src/spec/schema.rs

```diff
@@ -642,6 +644,204 @@ impl SchemaVisitor for IndexByName {
     }
 }
 
+struct PruneColumn {
+    selected: HashSet<i32>,
+    select_full_types: bool,
+}
+
+/// Visit a schema and returns only the fields selected by id set
+pub fn prune_columns(
+    schema: &Schema,
+    selected: HashSet<i32>,
+    select_full_types: bool,
+) -> Result<Option<Type>> {
+    let mut visitor = PruneColumn::new(selected, select_full_types);
+    visit_schema(schema, &mut visitor)
+}
+
+impl PruneColumn {
+    fn new(selected: HashSet<i32>, select_full_types: bool) -> Self {
+        Self {
+            selected,
+            select_full_types,
+        }
+    }
+
+    fn project_selected_struct(projected_field: Option<Type>) -> Result<StructType> {
+        match projected_field {
+            // If the field is a StructType, return it as such
+            Some(Type::Struct(s)) => Ok(s),
+            Some(_) => Err(Error::new(
+                ErrorKind::DataInvalid,
```

Comment:
```suggestion
                ErrorKind::Unexpected,
```
This should be a bug, and we should return `Unexpected` error.